src/metabase/lib/card.cljc

```diff
@@ -116,19 +116,27 @@
 (def ^:private CardColumns
   [:maybe [:sequential {:min 1} CardColumnMetadata]])
 
-(mu/defn ^:private card-metadata-columns :- CardColumns
+(def ^:private ^:dynamic *card-metadata-columns-card-ids*
+  "Used to track the ID of Cards we're resolving columns for, to avoid inifinte recursion for Cards that have circular
+  references between one another."
+  #{})
```

Comment:
Improvement: MLv2 should not barf when fetching metadata for Cards that have circular references to one another